NPE in Tradfri binding

From my logs, directly after discovering a light:
```
21:38:12.739 [ERROR] [alifornium.core.network.CoapEndpoint] - Exception in protocol stage thread: null
java.lang.NullPointerException
	at org.eclipse.smarthome.binding.tradfri.handler.TradfriLightHandler$LightData.getBrightness(TradfriLightHandler.java:260)[184:org.eclipse.smarthome.binding.tradfri:0.9.0.201706161831]
	at org.eclipse.smarthome.binding.tradfri.handler.TradfriLightHandler.onUpdate(TradfriLightHandler.java:122)[184:org.eclipse.smarthome.binding.tradfri:0.9.0.201706161831]
	at org.eclipse.smarthome.binding.tradfri.internal.TradfriCoapHandler.onLoad(TradfriCoapHandler.java:61)[184:org.eclipse.smarthome.binding.tradfri:0.9.0.201706161831]
	at org.eclipse.californium.core.CoapClient$ObserveMessageObserverImpl.deliver(CoapClient.java:1044)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.CoapClient$MessageObserverImpl.succeeded(CoapClient.java:970)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.CoapClient$MessageObserverImpl.onResponse(CoapClient.java:949)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.coap.Request.setResponse(Request.java:415)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.EndpointManager$ClientMessageDeliverer.deliverResponse(EndpointManager.java:249)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.CoapStack$StackTopAdapter.receiveResponse(CoapStack.java:193)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.ObserveLayer.receiveResponse(ObserveLayer.java:138)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.BlockwiseLayer.receiveResponse(BlockwiseLayer.java:318)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.ReliabilityLayer.receiveResponse(ReliabilityLayer.java:263)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.CoapStack.receiveResponse(CoapStack.java:135)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.receiveMessage(CoapEndpoint.java:630)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.access$700(CoapEndpoint.java:536)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl$1.run(CoapEndpoint.java:548)[182:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$5.run(CoapEndpoint.java:695)[182:californium-osgi:1.0.5]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_112]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_112]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_112]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_112]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_112]
```

Looks as if we need some more null safeguard...

I got this one today. Situation was a little bit different. My lamp was powerless for some days. After switching it on, the NPE was thrown. I have used the opportunity and created a PR #3784.
```
2017-07-01 12:19:20.944 [ERROR] [alifornium.core.network.CoapEndpoint] - Exception in protocol stage thread: null
java.lang.NullPointerException
	at org.eclipse.smarthome.binding.tradfri.handler.TradfriLightHandler$LightData.getBrightness(TradfriLightHandler.java:259)[216:org.eclipse.smarthome.binding.tradfri:0.9.0.b5]
	at org.eclipse.smarthome.binding.tradfri.handler.TradfriLightHandler.onUpdate(TradfriLightHandler.java:121)[216:org.eclipse.smarthome.binding.tradfri:0.9.0.b5]
	at org.eclipse.smarthome.binding.tradfri.internal.TradfriCoapHandler.onLoad(TradfriCoapHandler.java:60)[216:org.eclipse.smarthome.binding.tradfri:0.9.0.b5]
	at org.eclipse.californium.core.CoapClient$ObserveMessageObserverImpl.deliver(CoapClient.java:1044)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.CoapClient$MessageObserverImpl.succeeded(CoapClient.java:970)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.CoapClient$MessageObserverImpl.onResponse(CoapClient.java:949)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.coap.Request.setResponse(Request.java:415)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.EndpointManager$ClientMessageDeliverer.deliverResponse(EndpointManager.java:249)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.CoapStack$StackTopAdapter.receiveResponse(CoapStack.java:193)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.ObserveLayer.receiveResponse(ObserveLayer.java:138)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.BlockwiseLayer.receiveResponse(BlockwiseLayer.java:318)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.ReliabilityLayer.receiveResponse(ReliabilityLayer.java:263)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:98)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.stack.CoapStack.receiveResponse(CoapStack.java:135)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.receiveMessage(CoapEndpoint.java:630)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.access$700(CoapEndpoint.java:536)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl$1.run(CoapEndpoint.java:548)[214:californium-osgi:1.0.5]
	at org.eclipse.californium.core.network.CoapEndpoint$5.run(CoapEndpoint.java:695)[214:californium-osgi:1.0.5]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_65]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_65]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_65]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_65]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_65]
```

--- a/extensions/binding/org.eclipse.smarthome.binding.tradfri/README.md
+++ b/extensions/binding/org.eclipse.smarthome.binding.tradfri/README.md
@@ -21,7 +21,7 @@ The devices require only a single (integer) parameter, which is their instance i
 
 ## Channels
 
-All devices support the `brightness` channel, while the white spectrum bulbs additionally also support the `colorTemperature` channel.
+All devices support the `brightness` channel, while the white spectrum bulbs additionally also support the `color_temperature` channel.
 
 | Channel Type ID   | Item Type | Description                                 |
 |-------------------|-----------|---------------------------------------------|
@@ -51,7 +51,7 @@ demo.sitemap:
 sitemap demo label="Main Menu"
 {
     Frame {
-    	Slider item=Light label="Brightness [%.1f %%]"
-	}
+        Slider item=Light label="Brightness [%.1f %%]"
+    }
 }
 ```

--- a/extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriLightHandler.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriLightHandler.java
@@ -256,11 +256,16 @@ public class TradfriLightHandler extends BaseThingHandler implements CoapCallbac
         }
 
         public PercentType getBrightness() {
-            int b = attributes.get(DIMMER).getAsInt();
-            if (b == 1) {
-                return new PercentType(1);
+            JsonElement dimmer = attributes.get(DIMMER);
+            if (dimmer != null) {
+                int b = dimmer.getAsInt();
+                if (b == 1) {
+                    return new PercentType(1);
+                }
+                return new PercentType((int) Math.round(b / 2.54));
+            } else {
+                return null;
             }
-            return new PercentType((int) Math.round(b / 2.54));
         }
 
         public boolean getReachabilityStatus() {
@@ -278,7 +283,12 @@ public class TradfriLightHandler extends BaseThingHandler implements CoapCallbac
 
         @SuppressWarnings("unused")
         public int getTransitionTime() {
-            return attributes.get(TRANSITION_TIME).getAsInt();
+            JsonElement transitionTime = attributes.get(TRANSITION_TIME);
+            if (transitionTime != null) {
+                return transitionTime.getAsInt();
+            } else {
+                return 0;
+            }
         }
 
         public LightData setColorTemperature(PercentType c) {
@@ -307,7 +317,7 @@ public class TradfriLightHandler extends BaseThingHandler implements CoapCallbac
             // we only need to check one of the coordinates and figure out where between the presets we are
             JsonElement colorX = attributes.get(COLOR_X);
             if (colorX != null) {
-                double x = attributes.get(COLOR_X).getAsInt();
+                double x = colorX.getAsInt();
                 double value = 0.0;
                 if (x > X[1]) {
                     // is it between preset 1 and 2?

--- a/extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/internal/discovery/TradfriDiscoveryService.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/internal/discovery/TradfriDiscoveryService.java
@@ -37,7 +37,8 @@ public class TradfriDiscoveryService extends AbstractDiscoveryService implements
 
     private TradfriGatewayHandler handler;
 
-    private String[] COLOR_TEMP_MODELS = new String[] { "TRADFRI bulb E27 WS opal 980lm" };
+    private String[] COLOR_TEMP_MODELS = new String[] { "TRADFRI bulb E27 WS opal 980lm",
+            "TRADFRI bulb GU10 WS 400lm" };
 
     public TradfriDiscoveryService(TradfriGatewayHandler bridgeHandler) {
         super(TradfriBindingConstants.SUPPORTED_LIGHT_TYPES_UIDS, 10, true);

